execve失敗時の終了ステータスを修正

nafuka11 opened the pull request:
Fix #77

## やったこと
1. execveで `errno` が `ENOEXEC` の場合、エラーメッセージを表示せず0でexitするよう修正。
   - 例：実行フォーマットでないファイル（空ファイルなど）が渡された時。
2. execveで `ENOENT` 以外の `errno`だった場合126でexitするよう修正。
   - 例：ファイルがディレクトリ扱いの時（ `Not a directory` ）

### テストケース
ケース | 説明
-- | --
/bin/ls/no_such_file | 絶対パスでファイルがディレクトリとして指定されている場合。
/bin/no_such_dir/file | 絶対パスでディレクトリが存在しない場合。
./empty, touch empty; chmod +x empty | 実行フォーマットでないファイルが指定された場合。

nafuka11 committed:
fix(exec): execve失敗時ENOENT以外は終了ステータス1でなく126とするよう修正

nafuka11 committed:
fix(exec): execveのerrnoがENOEXECの時正常終了扱いとするよう修正

エラーメッセージを表示せず126でなく0を返すように

nafuka11 committed:
test(simple_command): execveに失敗するケースを追加

1. 絶対パスでディレクトリは存在、ファイルは存在しない時
2. 絶対パスでディレクトリ、ファイルが存在しない時
3. 実行可能なフォーマットでないファイルを指定した場合

ToYeah reviewed (commented):
bashの仕様確認と修正ありがとうございます。問題ないと思います。

`execve`側でエラーを出せるようになったので
https://github.com/ToYeah/42_minishell/blob/f3bf2c854588d866f632a9449ffe4d73d1ce2d1b/srcs/exec/build_path.c#L56
ここの`126`が返る二つの`if`文なくても動くと思います！(消してみてテスト通る事確認しました)

これは僕のミスなんですけど、`stat`が失敗したのにTRUEが返ってしまってる部分もついでに直してもらえると嬉しいです！！！！
https://github.com/ToYeah/42_minishell/blob/f3bf2c854588d866f632a9449ffe4d73d1ce2d1b/srcs/exec/build_path_stat.c#L21

nafuka11 commented:
すいません、テスト時に setupスクリプトが動いておらず、きちんとした動作になっていませんでした……。

is_directoryのif文は消してしまうと、 `is a directory` が出力されなくなってしまうのでこれは必要……なのですが、bashだと `execve` 失敗後にディレクトリか判定してそうなので、処理を見直そうと思います。

nafuka11 committed:
test(grademe): prepare_test_dirでsetupスクリプトを実行するよう修正

fix #83

nafuka11 committed:
test(grademe): test実行前にmakeするように

nafuka11 committed:
feat(exec): handle_execve_error追加

execveに失敗した時に呼び出される
- pathがディレクトリの時は "is a directory"
- ENOEXEC かつ 実行不可の時 "Permission denied" のerrnoを設定（300などread権限がない時）
- ENOEXEC の時は正常終了扱い。それ以外はエラーを表示して終了する

nafuka11 committed:
fix(exec): is_executableの判定修正

- pathが存在しない（stat==-1）時FALSEを返すように
- read権限がない時FALSEを返すように

nafuka11 committed:
fix(exec): check_cmd_pathではpathのNULL判定だけ行うように

nafuka11 committed:
chore(exec): exec_binaryでhandle_execve_errorを呼び出すように

nafuka11 committed:
test(simple_command): read権限がない時、ある時のテストケース追加

./permでなくpermのテストはexportを実装してから行います

nafuka11 committed:
Merge branch 'master' into 77-fix-exit-status-when-execve-failed

nafuka11 commented:
## 修正内容
1. is_executableの判定修正
   - ご指摘の、statが-1の時FALSEを返すようにしました。
   - また、read権限がない時もFALSEを返すようにしています。
2. check_cmd_pathではpathがNULLかどうかだけ判定するようにしました。
   - ディレクトリ判定、executable判定は、新規作成した `handle_execve_error` で行なっています。
3. execveのエラー処理関数、 `handle_execve_error` の追加。
   - 最初に終了ステータスの設定。
   - 次にpathがディレクトリか判定。ディレクトリなら `is a directory` を表示して `exit`。
   - executableでない かつ errnoが `ENOEXEC` の場合、errnoを `EACCES` に設定（read権限がないファイルで`Exec format error` でなく `Permission denied` と表示させるための処理です）
   - それでもerrnoが `ENOEXEC` なら、正常終了扱い。
   - それ以外なら、`print_error` して `exit` します。
4. テストの追加
   - ファイルにread権限がある時、ない時のテストを追加しました。
   - コマンド名のみ（./perm でなく perm）のテストは、exportを実装したらPATHを設定する予定です。

ToYeah commented on srcs/exec/command.c:
```diff
@@ -98,3 +101,24 @@ void		wait_commands(t_command *command)
+	}
+	if (errno == ENOEXEC && !is_executable(path))
+		errno = EACCES;
+	if (errno == ENOEXEC)
+		exit(EXIT_SUCCESS);
```
execveは失敗しているのに成功扱いにするときってどんな時ですか？

nafuka11 replied:
空ファイルに実行権限がついているケースです。
```bash
bash-3.2$ touch empty
bash-3.2$ chmod +x empty
bash-3.2$ ./empty
bash-3.2$ echo $?
0
```

ToYeah replied:
なるほど!!
そこに効いてくるんですね!
